### DISCUSSION TITLE
모든 테스트 코드를 해당 로직만 테스트하도록 전환하자

### DISCUSSION THREAD

**Discussion by kyum-q** [woowacourse-teams/2024-code-zap#760] [이야기해봐요]
현재 테스트 코드를 보면, 테스트 대상 객체의 비즈니스 로직 외에도 해당 객체가 사용하는 다른 객체들의 비즈니스 로직까지 함께 테스트하는 방식으로 작성되어 있습니다. 이를 **테스트 대상 객체의 비즈니스 로직만 테스트하도록 변경** 하는 것이 어떨까요?

편의를 위해 용어를 정리하면,
**화이트박스 테스트** : 객체의 내부 구현과 로직을 알고 그 세부 동작까지 검증하는 방식을 의미
**블랙박스 테스트** : 객체의 비즈니스 로직만을 대상으로 내부 동작을 구체적으로 알지 못한 채, 외부에서 입력과 출력만을 검증하는 방식을 의미
예를 들어, 컨트롤러 테스트 시 서비스 클래스의 내부 로직을 알지 않고, **메서드 시그니처(메서드 명, 입력값, 반환값)** 만 보고 테스트를 작성하는 것은 블랙박스 테스트에 가깝습니다.

따라서, 현재 테스트를 **블랙박스 테스트 방식** 으로 구현하는 것이 더 나은 방향이 될 것 같습니다. 그 이유는 다음과 같습니다.

> 하지만 완전한 블랙박스 테스트를 말하는 건 아님!

### 1. 중복된 테스트 코드 문제

현재 방식에서는 하위 클래스에 대한 검증이나 로직이 추가될 때마다, 상위 클래스 테스트에도 관련된 테스트를 작성해야 하는 번거로움이 있습니다.
이로 인해 상위 테스트가 하위 테스트의 역할까지 포함하게 되며, 결국 하위 테스트의 의미가 퇴색된다고 생각합니다.
상위 테스트가 화이트박스 테스트 방식으로 모든 예외 처리를 포함한다면, 하위 테스트의 필요성이 줄어듭니다.
또한, 중복된 테스트 코드로 인해 테스트 개수가 불필요하게 많아지고, 테스트 실행 시간도 길어질 수 있습니다.

### 2. 테스트 코드 가독성 저하

예를 들어, 템플릿 파사드 서비스에서 템플릿 삭제 로직을 테스트할 때, 화이트박스 테스트 방식으로 모든 로직을 테스트하면, 중복된 테스트 코드로 인해 특정 검증을 찾기 어려워집니다.
예를 들어, “중간에 에러가 발생했을 때 롤백되는지”와 같은 로직을 검증할 때, 테스트 코드에서 해당 로직을 바로 확인하기 어려워질 수 있습니다.

### 3. 기타 문제

현재 fake DB를 사용하고 있으니 이를 제거하고, fake DB를 사용하는 테스트를 모두 실제 DB를 사용하도록 변경하는 것도 어려움이 있습니다.
특히, 현재 컨트롤러 테스트는 서비스 로직까지 모두 포함하고 있어, 말만 컨트롤러 테스트일 뿐 실제로는 서비스 로직까지 과도하게 테스트하고 있습니다.

## 결론

따라서 모든 테스트를 현재 **테스트 대상 객체만 신경 쓰는 블랙박스 테스트 방식으로 변경** 하여, 테스트 간의 역할을 명확히 분리하고, 중복된 테스트를 제거하는 것이 더 효율적일 것 같습니다.

**Comment by zeus6768** [comment]
100% 동의해요~ 컨트롤러에서 서비스나 레포지토리의 로직까지 검증해줄 이유가 없다고 생각합니다.

**Comment by zangsu** [comment]
저도 좋다고 생각해요!
다만 아직까지 블랙박스 테스트를 **어떻게** 해야할 지 잘 감이 안잡혀서,,,
Mockito 등의 모킹을 사용하는 것이 되려나용 👀

**Reply by zangsu** [reply]
단위테스트 넘모 어려웡

**Comment by kyum-q** [comment]
각 계층에서 테스트 해야 하는 것을 명확하게 하자 ~!

### 컨트롤러 테스트
- 성공 확인 (request에 따른 적절한 response가 반환되는지)
- dto 예외 처리

> 모킹을 사용하자

### 서비스 테스트
- 파라미터에 따른 반환 값 확인
- 서비스 단에서 하는 예외 처리

### 레파지토리 테스트
- 파라미터에 따른 반환 값 확인
   - DB에 잘 넣어졌냐?
   - 새로 만든 메서드 잘 작동?

### 도메인 테스트
- 파라미터에 따른 반환 값 확인
- 도메인이 하는 일을 잘 하냐?
